src/compile/data/bin.ts

```diff
@@ -1,110 +1,141 @@
-import {DataComponentCompiler} from './base';
-
-import {Bin, binToString} from '../../bin';
+import {autoMaxBins, Bin, binToString} from '../../bin';
 import {Channel} from '../../channel';
 import {field, FieldDef} from '../../fielddef';
 import {hasDiscreteDomain} from '../../scale';
-import {Dict, extend, flatten, isBoolean, vals, varName} from '../../util';
+import {Dict, extend, flatten, hash, isBoolean, StringSet, vals, varName} from '../../util';
 import {VgBinTransform, VgTransform} from '../../vega.schema';
-
-import {FacetModel} from './../facet';
-import {LayerModel} from './../layer';
 import {Model} from './../model';
+import {DataFlowNode, DependsOnNode, NewFieldNode} from './dataflow';
 
 
 function numberFormatExpr(expr: string, format: string) {
   return `format(${expr}, '${format}')`;
 }
 
-function addRangeFormula(model: Model, transform: VgTransform[], fieldDef: FieldDef, channel: Channel) {
+function rangeFormula(model: Model, fieldDef: FieldDef, channel: Channel) {
     const discreteDomain = hasDiscreteDomain(model.scale(channel).type);
-    if  (discreteDomain) {
+
+    if (discreteDomain) {
       // read format from axis or legend, if there is no format then use config.numberFormat
       const format = (model.axis(channel) || model.legend(channel) || {}).format ||
         model.config.numberFormat;
 
       const startField = field(fieldDef, {datum: true, binSuffix: 'start'});
       const endField = field(fieldDef, {datum: true, binSuffix: 'end'});
 
-      transform.push({
-        type: 'formula',
-        as: field(fieldDef, {binSuffix: 'range'}),
-        expr: `${numberFormatExpr(startField, format)} + ' - ' + ${numberFormatExpr(endField, format)}`
-      });
+      return {
+        formulaAs: field(fieldDef, {binSuffix: 'range'}),
+        formula: `${numberFormatExpr(startField, format)} + ' - ' + ${numberFormatExpr(endField, format)}`
+      };
     }
+    return {};
+}
+
+interface Component {
+  bin: Bin;
+  field: string;
+  extentSignal: string;
+  signal: string;
+  as: string[];
+  /**
+   * Range formula.
+   */
+  formula?: string;
```

Comment:
I think I miss one of the other parameter required, but let's not worry about this for now as it is minor.